File: 프로그래머스/lv1/42889. 실패율/실패율.py
def solution(N, stages):
    answer = []
    stagePeopleCnt = [0] * (N + 2)
    
    for i in range(len(stages)):
        stagePeopleCnt[stages[i]] += 1
        
    people = stagePeopleCnt[N + 1]
    for i in range(N, 0, -1):
        people += stagePeopleCnt[i]
        failRate = stagePeopleCnt[i] / people if people != 0 else 0
        answer.append([i, failRate])
        
    answer = sorted(answer, key=lambda x : (-x[1], x[0]))
    
    return [arr[0] for arr in answer]

File: 프로그래머스/lv1/42889. 실패율/실패율.java
import java.util.*;

class Solution {
    public int[] solution(int N, int[] stages) {
        int[] answer = new int[N];
        int[] notClear = new int[N + 2];
        double[][] failRates = new double[N][2];
        
        // 스테이지 별 도전자 수 구하기
        for (int i = 0; i < stages.length; i++)
            ++notClear[stages[i]];
        
        // 스테이지 실패율 구하기
        int people = notClear[N + 1]; // 도달자 수
        for (int i = N; i > 0; i--) {
            people += notClear[i];
            failRates[i - 1][0] = i;
            failRates[i - 1][1] = people != 0 ? (double) notClear[i] / people : 0;
        }
        
        // 정렬
        Arrays.sort(failRates, (f1, f2) -> Double.compare(-f1[1], -f2[1]));
        
        // 스테이지 넘버만 뽑기
        for (int i = 0; i < N; i++)
            answer[i] = (int) failRates[i][0];
        
        return answer;
    }
}